megatron/data/indexed_dataset.py

```
@@ -1065,3 +1065,27 @@ def gather_files_dist(filemain, filelist, distctx):
         gather_files_dist_idx_cached(filemain, filelist, distctx)
     elif indexstr == "mmap":
         gather_files_dist_idx_mmap(filemain, filelist, distctx)
+
+
+def get_start_end(count, rank, numranks):
+    num, remainder = divmod(count, numranks)
+    if rank < remainder:
+        start = (num + 1) * rank
+        end = start + num + 1
+    else:
+        start = (num + 1) * remainder + num * (rank - remainder)
+        end = start + num
+    return start, end
```

doesn't this code already exist? Ah sorry it was in another PR, nvm. https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/60/files#diff-d4fb2ffb288c70937ba07e1e59497a36897269ea68b94985462dc51bcf2c392aR90-R103

---

So I'm pretty sure this still works for `count < numranks`, but maybe we should throw an exception in that case. This is interesting, because in our current env, we have to request a certain amount of allocation, and asking to much is just a waste IMO. (Btw the exception could be outside this method)

---

Good idea, but maybe print a warning instead of throw an exception?  The merge is pretty fast, so it's a bit more user friendly to just go ahead and do it, even if some ranks have no work to do.  The warning will help inform the user to kill the run or use less resources the next time.

---

Added a warning.